### ISSUE TITLE
issue with MathJax Latin Modern fonts on Firefox using native MathML output [was Upgrade latin modern math to 1.959]

### ISSUE THREAD

**Issue by flying-sheep** [mathjax/MathJax#1667]
see https://bugs.launchpad.net/ubuntu/+source/lmodern/+bug/1502701

**Comment by pkra** [comment]
Thanks for filing this.

Their README seems to only contain a comment on the changes. Those indicate only one relevant change (fixing contour integrals positions); we could fix that without regenerating the fonts if we wanted.

**Comment by flying-sheep** [comment]
it says:

> Changes were made to the HHEA and OS/2 tables (all fonts) and the contour integrals (Pagella, Termes, Latin Modern).

and the important bugfix is the OS/2 thing. without it, firefox displays everything as if the line height was super tall

**Comment by dpvc** [comment]
MathJax repackaged the fonts for its internal use, and I don't believe that the HHEA and OS/2 tables are taken from the original, but rather are created anew.  Also, MathJax's HTML-CSS output does its own positioning and clipping, so the line height would not affect the surroundings even if it were.  The SVG output uses font paths, not the font itself, so the change is immaterial.  Finally, the CommonHTML output doesn't support Latin Modern at the moment.  So I don't think there should be any issue with the current version of the Latin Modern fonts that MathJax uses.

**Comment by flying-sheep** [comment]
i had the described problem because my system chose to use mathjax’ version of Latin Modern Math from https://github.com/mathjax/MathJax/archive/2.7.0.tar.gz (fontforge says it’s 1.958)

this means the font’s OS/2 table packaged there has the bug.

so either the way that table is created has to be fixed or the source font has to be upgraded.

**Comment by dpvc** [comment]
I'm a little confused as the archive you mention doesn't include `Latin Modern Math`.  All the MathJax Latin Modern fonts should be named `LatinModernMathJax_` followed by something like `Main` or `Monospace` or any of 2 dozen other names.  But none of them is `Math`.  The internal names should be "Latin Modern MathJax Main", etc., if you are looking at them in FontForge.

I'm not sure what you mean by "your system chose" MathJax's versions, since these are non-standard names and should not be ones used by other applications.  They are not intended for use by other applications, and the font has been broken up into pieces that are only useful to MathJax.  Can you be more specific about what is actually happening?

The only thing I can think of is that "Latin Modern MathJax Main" is being truncated by your system to "Latin Modern Math" for some reason (but since that is 17 characters, that doesn't seem likely).

**Comment by dpvc** [comment]
> this means the font’s OS/2 table in that font has the bug.

For its use in MathJax that doesn't matter.  The MathJax fonts are not meant to be system fonts used in other ways, so if your system is picking them up for use outside of MathJax, perhaps you can block the MathJax directory from your font system?

**Comment by pkra** [comment]
> and the important bugfix is the OS/2 thing. without it, firefox displays everything as if the line height was super tall

I'm guessing you use the NativeMML output and somehow force Firefox to use  `LatinModernMathJax_xxx` fonts; neither is advisable from a MathJax point of view. But you should be able to install a font that Firefox supports (e.g., the original Latin Modern fonts) and then configure Firefo to force that font for MathML content (there's a configuration option nowadays).

**Comment by flying-sheep** [comment]
the internal name is “Latin Modern Math”:

```console
$ fc-match 'Latin Modern Math'
LatinModernMathJax_Alphabets-Regular.woff: "Latin Modern Math" "Regular"
```

disabling “Latin Modern Math” with KDE’s systemsettings weirdly renames all font files so that they start with a dot and adds a `disabledfonts.xml` with this content:

```xml
<disabledfonts>
  <font family="Latin Modern Math" weight="80" width="100" slant="0" >
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Arrows-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Normal-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_NonUnicode-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Size5-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Latin-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Shapes-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_DoubleStruck-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Size3-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Operators-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Monospace-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Size2-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Size1-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Alphabets-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Size7-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Misc-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Script-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Size6-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Size4-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Variants-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Symbols-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Fraktur-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Main-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_Marks-Regular.woff" foundry="PfEd"/>
   <file path="/usr/share/fonts/mathjax/HTML-CSS/Latin-Modern/woff/.LatinModernMathJax_SansSerif-Regular.woff" foundry="PfEd"/>
  </font>
</disabledfonts>
```

**Comment by pkra** [comment]
> I did nothing of the sort. 

Sorry, I didn't mean to imply any intention, just that your system configuration ends up doing what it's doing. 

Thanks for the additional data on the family; we'll check what we can do. As @dpvc wrote, blocking those MathJax fonts would be the sensible.

And as I wrote, you should be able to install the original Latin Modern fonts (they're on debian/ubuntu) and set the preference in Firefox to use them for MathML.

**Comment by flying-sheep** [comment]
I apologize too, that first sentence was too harsh! I wanted to say that I didn’t force firefox to use the font, I guess it’s just used because of its “Preferred Family” being in a font stack used for MathML rendering.

I don’t know how to priorize fonts of the same name, especially not from firefox. Firefox asks fontconfig, fontconfig probably just uses whatever if there’s no rule telling it otherwise (so I can just use a rule to block those files to work around this bug)

**Comment by pkra** [comment]
> I don’t know how to priorize fonts of the same name, especially not from firefox. 

Check Preferences => Content => Fonts advanced => "fonts for" => Math

> I’d say this isn’t a [Question], right?

On the one hand, I've filed a bug in the dev-tools repo regarding font updates. On the other hand, your issue is unrelated to MathJax itself -- it can't control what fonts are installed on a system, let alone what fonts Firefox uses with its native MathML rendering. That is why we advise not to use MathML output anymore -- the results are unpredictable and outside our control.

**Comment by flying-sheep** [comment]
Hi, since I had to circumvent this again on another system, I came across this bug again.

It might be unrelated to MathJax-the-tool, but it’s an issue with a patched font you ship in this repo, which is clearly bugged in two ways:

1. It is an old version with bugged metrics that can be fixed by upgrading it
2. Your renaming procedure was incomplete. The “Preferred Family” TTF name is still “Latin Modern Math”

It would be nice to fix one or both of these issues.

In 337361c1cd4c9465385bc267a7b14dcc09ab0a53/#611 you “regenerated the fonts”. What does that mean? Is there a script somewhere?
